### PR TITLE
feat(rule): add prefer-output-readonly rule

### LINKED ISSUE #515: [Feature Request] - Force output fields to be readonly
I'm not sure if this is 100% correct, since I don't recall seeing it anywhere else, but in our Angular project, we are marking all `@Output` fields `readonly` because really, there is no need to ever re-assign them.

Unless there are some edge cases, which we aren't aware of right now, it would be great if codelyzer could do the check, if all `@Output` fields are marked as `readonly` for us, otherwise we always need to check for it manually in code reviews.

Similarly it could be also checked that `@Input` is never marked as `readonly` (I have seen it sometimes in our code base and typescript never complained about it, even though it was definitely wrong and set at a later point).

I've never implemented a tslint rule myself, but if you think this one makes sense and you know of some which I could basically copy/paste with few adjustments, to achieve those checks, I'd be fine with contributing the rule as well.

### PR DESCRIPTION
**Pull request by rafaelss95**
Closes #515.

### CODE REVIEW

File: src/preferOutputReadonlyRule.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+export class OutputMetadataWalker extends NgWalker {
+  visitNgOutput(property: ts.PropertyDeclaration, output: ts.Decorator, args: string[]) {
+    if (property.modifiers && property.modifiers.some(m => m.kind === ts.SyntaxKind.ReadonlyKeyword)) {
+      return;
+    }
+
+    const className = (property.parent as ts.PropertyAccessExpression).name.getText();
+    const memberName = property.name.getText();
+    this.addFailureAtNode(property.name, Rule.FAILURE_STRING);
```
[the remark this answers is not in the document]
Comment (rafaelss95) [reply src/preferOutputReadonlyRule.ts]: I've fixed it. Btw, none of the four occurrences that I found, propagate the call `super.visitNgOutput`.

File: src/preferOutputReadonlyRule.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  }
+}
+
+export class OutputMetadataWalker extends NgWalker {
+  visitNgOutput(property: ts.PropertyDeclaration, output: ts.Decorator, args: string[]) {
+    if (property.modifiers && property.modifiers.some(m => m.kind === ts.SyntaxKind.ReadonlyKeyword)) {
+      return;
+    }
+
+    const className = (property.parent as ts.PropertyAccessExpression).name.getText();
```
[the remark this answers is not in the document]
Comment (rafaelss95) [reply src/preferOutputReadonlyRule.ts]: In all occurrences that I found, we never check if `parent` is `null`.

File: src/preferOutputReadonlyRule.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import * as Lint from 'tslint';
+import * as ts from 'typescript';
+import { NgWalker } from './angular/ngWalker';
+
+export class Rule extends Lint.Rules.AbstractRule {
+  public static metadata: Lint.IRuleMetadata = {
+    description: 'Prefer to declare `@Output` as readonly since they are not supposed to be reassigned.',
+    options: null,
+    optionsDescription: 'Not configurable.',
+    rationale: '',
```
Comment (rafaelss95) [inline src/preferOutputReadonlyRule.ts]: A rationale? :)

File: src/preferOutputReadonlyRule.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgechev) [reply src/preferOutputReadonlyRule.ts]: Then we should fix the rest :-)

**Pull request merged**